blog/2024-11-18-relay-ops.md

```diff
@@ -0,0 +1,60 @@
+
+Best-effort consumers can likely be configure to consume from the current firehose offset, and simply restart, missing at most a few seconds of events. This probably makes sense for feed generators, analytics, automated labeling, etc.
+
+Services which want to really minimize the number of missed events can take a different approach. They can be programmed to connect to the new relay with no cursor, detect the current sequence number, then reconnect with a cursor a couple minutes back (at current rates this means a few hundred thousand events back). This results in a time overlap window between the two relays (with some events processed twice), which helps minimize the chance of missing any events.
+
+## Firehose Fan-out and Rainbow
+
+Firehose fan-out servers have a single upstream connection to a relay firehose, and re-broadcast that stream to multiple subscribing clients. They maintain a local backfill window (allowing re-connection with a cursor), but do not implement endpoints like `getRecord` or `getRepo`, only `com.atproto.sync.subscribeRepos`. They do not re-validate the event stream (eg, don't verify signatures). They maintain the exact sequence numbers of the upstream relay, meaning that they are interchangeable with the relay and with sibling fan-out servers, and can be placed behind a load-balancer.
+
+Today we are releasing Rainbow, a fan-out service implemented in Go. The [source code is available](https://github.com/bluesky-social/indigo/tree/main/cmd/rainbow) in the `indigo` git repository, with the same MIT/Apache licensing as our other open source projects.
+
+In the future, we may run additional Rainbow instances outside our primary data centers, similar to how we offer Jetstream instances today. This distributes the networking over more routers, and could improve connection reliability and throughput for subscribers in regions outside North America.
+
+Going forward, we recommend that software which consumes from firehoses (including atproto SDKs) support HTTP redirects for WebSocket connections. This enables "pooling" behavior where a single hostname could route clients to multiple distinct servers, without use of a load-balancer.
+
+## What Else?
+
+As a reminder, the Relay synchronization API is the same as the PDS synchronization API, including the firehose. Relays are essential to the functionality of the atproto network: they are an operational and developer convenience, allowing consuming services to skip keeping track of which PDS instances are active.
+
+A classic way to scale services like the firehose is *sharding*, where the stream is split in to multiple parallel streams. atproto has a natural sharding key (the account DID), which means related events can be consistently routed to the correct shard. This will allow near-indefinite scaling of the firehose event rate, assuming compute and network resources are available.
```

Comment:
```suggestion
A classic way to scale services like the firehose is *sharding*, where the stream is split into multiple parallel streams. atproto has a natural sharding key (the account DID), which means related events can be consistently routed to the correct shard. This will allow near-indefinite scaling of the firehose event rate, assuming compute and network resources are available.
```
`near-indefinite scaling` or `near-infinite scaling`, I'm not sure which would be clearer here? 🤔